### DISCUSSION TITLE
BentoML Roadmap Update - Q4 2020

### DISCUSSION THREAD

**Discussion by parano** [bentoml/BentoML#1128] [General]
Hi everyone, this is a snapshot of our current roadmap for BentoML. We are going back to the schedule of releasing a new version every 2 weeks in October, release 0.9.0 will come out this week, and here's what to expect in the next a few releases. Comment below if you are interested in some of the features listed here, that will help us prioritize. And feel free to suggest other features you'd like to see in future versions of BentoML.

Edit: We have moved all roadmap information to the GitHub Projects page, under the `Roadmap` project: https://github.com/bentoml/BentoML/projects/1

**Comment by parano** [comment]
#### Improve Model Management functionality
   * **0.10.0** Add the push/pull API for moving saved BentoService to and from a remote YataiService
   * **0.10.0** Add documentation on best practices for running a shared YataiService for team collaboration
   * **0.11.0** Add YataiService's own monitoring endpoint #1060
   * Allow configuring YataiService web UI's prefix URL #1063 
   * Allow attaching additional metadata with packaged model #881

**Comment by parano** [comment]
#### Improve Documentation

   * **0.10.0** Best practices on integrating with model development workflow and training pipelines
   * **0.10.0** Example projects showing sample project structure and best practices
   * **0.11.0** Deep dive on how BentoML's adaptive micro-batching mechanism works and how it compares to Clipper and TF-serving's implementation
   * **0.12.0** Finish the rest of the advanced guides section, on monitoring, logging best practices, how to customize artifact and adapters

**Comment by parano** [comment]
#### More built-in Input/Output adapters

  * **0.10.0** Add FileOutput adapter
  * **0.10.0** Add ImageOutput adapter
  * **0.11.0** Add numpy ndarray input/output adapter

**Comment by parano** [comment]
#### Better Spark Support
   *  **0.11.0** Add Spark MLLib support #666
   *  **0.11.0** Document best practices on running distributed inference job with BentoML packaged model on Spark #890
   *  **0.12.0** Add type convertion in defaul t input/output adapter for working with Spark dataset

**Reply by Talador12** [reply]
This is probably the # 2 value addition in my eyes

**Comment by parano** [comment]
#### "One-click deployment" for more platforms
BentoML currently provides simple model deployment management for AWS Lambda, AWS SageMaker, and Azure Functions. And we want to add support for these additional platforms:

* **DONE**  **0.12.0**  Add support for deploying directly to AWS EC2 with auto-scaling and load balancer
* Add support for deploying directly to GCP Cloud Run
* Support micro-batching when deploying to AWS SageMaker #673

**Reply by mayurnewase** [reply]
I can take on this for aws with cloudformation templates for one click deployment.

**Reply by parano** [reply]
@mayurnewase that'd be great!! Let me know if you want to chat about the deployment operator design in BentoML or have any questions, sorry that this part of the project is still lacking proper documentation.  Ping me in slack if you need any help!

**Comment by parano** [comment]
#### API Server lifecycle callback functions
  *  **0.10.0** Add server startup and shutdown callback #1065 #1037 


#### GRPC support in API model server 
  *  **0.12.0** Add support for gRPC/protobuf data for all the built-in input/output adapters and allow exposing a gRPC serving endpoint
  * Support KFServing API V2 predict protocol https://github.com/bentoml/BentoML/issues/899


#### Better KFServing integration
  * **0.13.0** Support KFServing V2 API #899


#### Model Orchestration
For users that are managing model serving workloads on a Kubernetes cluster, we are developing advanced model orchestration tools for you, more details/design will be released soon
  * **0.12.0** Design doc & documentation for community review

**Reply by Talador12** [reply]
This is probably the # 1 value addition in my eyes

**Comment by parano** [comment]
#### Framework specific integration & support

* Better Tensorflow Integration https://github.com/bentoml/BentoML/issues/1188
* Add Pytorch Lightning support https://github.com/bentoml/BentoML/issues/1137
* PytorchModelArtifact support JIT traced models https://github.com/bentoml/BentoML/issues/873
* Pycaret support https://github.com/bentoml/BentoML/issues/988

**Comment by parano** [comment]
#### API server improvements

* Better GPU support https://github.com/bentoml/BentoML/issues/886
* Option to disable swagger UI https://github.com/bentoml/BentoML/issues/1167
* Allow adding extra custom URL routes https://github.com/bentoml/BentoML/issues/1161
* Allow configuring API server logging behavior https://github.com/bentoml/BentoML/issues/1009
* Allow setting micro-batching parameters from CLI argument https://github.com/bentoml/BentoML/issues/930

**Comment by parano** [comment]
#### Advanced Model Packaging features

* Allow specifying allowed Python version range https://github.com/bentoml/BentoML/issues/1004

**Comment by parano** [comment]
#### Testing

* Test dockerized API Server in artifact integration test 
 https://github.com/bentoml/BentoML/issues/891

**Comment by parano** [comment]
For those interested in following our progress on this roadmap, please refer to our roadmap project kanban here: https://github.com/bentoml/BentoML/projects/1
